File: gremlin-dotnet/src/Gremlin.Net/Structure/IO/GraphBinary/Types/EdgeSerializer.cs
#region License

/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

#endregion

using System;
using System.Collections.Generic;
using System.IO;
using System.Threading;
using System.Threading.Tasks;

namespace Gremlin.Net.Structure.IO.GraphBinary.Types
{
    /// <summary>
    /// An <see cref="Edge"/> serializer.
    /// </summary>
    public class EdgeSerializer : SimpleTypeSerializer<Edge>
    {
        /// <summary>
        ///     Initializes a new instance of the <see cref="EdgeSerializer" /> class.
        /// </summary>
        public EdgeSerializer() : base(DataType.Edge)
        {
        }

        /// <inheritdoc />
        protected override async Task WriteValueAsync(Edge value, Stream stream, GraphBinaryWriter writer,
            CancellationToken cancellationToken = default)
        {
            await writer.WriteAsync(value.Id, stream, cancellationToken).ConfigureAwait(false);
            // wrapping label into list here for now according to GraphBinaryV4
            await writer.WriteNonNullableValueAsync(new List<object> { value.Label }, stream, cancellationToken).ConfigureAwait(false);

            await writer.WriteAsync(value.InV.Id, stream, cancellationToken).ConfigureAwait(false);
            // wrapping label into list here for now according to GraphBinaryV4
            await writer.WriteNonNullableValueAsync(new List<object> { value.InV.Label }, stream, cancellationToken).ConfigureAwait(false);
            await writer.WriteAsync(value.OutV.Id, stream, cancellationToken).ConfigureAwait(false);
            // wrapping label into list here for now according to GraphBinaryV4
            await writer.WriteNonNullableValueAsync(new List<object> { value.OutV.Label }, stream, cancellationToken).ConfigureAwait(false);

            // Placeholder for the parent vertex
            await writer.WriteAsync(null, stream, cancellationToken).ConfigureAwait(false);

            // placeholder for the properties
            await writer.WriteAsync(null, stream, cancellationToken).ConfigureAwait(false);
        }

        /// <inheritdoc />
        protected override async Task<Edge> ReadValueAsync(Stream stream, GraphBinaryReader reader,
            CancellationToken cancellationToken = default)
        {
            var id = await reader.ReadAsync(stream, cancellationToken).ConfigureAwait(false);
            // reading single string value for now according to GraphBinaryV4
            var labelList = (List<object?>)await reader.ReadNonNullableValueAsync<List<object>>(stream, cancellationToken)
                .ConfigureAwait(false);
            var label = (string)labelList[0]!;

            var inVId = await reader.ReadAsync(stream, cancellationToken).ConfigureAwait(false);
            // reading single string value for now according to GraphBinaryV4
            var inVLabelList = (List<object?>)await reader.ReadNonNullableValueAsync<List<object>>(stream, cancellationToken).ConfigureAwait(false);
            var inV = new Vertex(inVId, (string)inVLabelList[0]!);

            var outVId = await reader.ReadAsync(stream, cancellationToken).ConfigureAwait(false);
            // reading single string value for now according to GraphBinaryV4
            var outVLabelList = (List<object?>)await reader.ReadNonNullableValueAsync<List<object>>(stream, cancellationToken).ConfigureAwait(false);
            var outV = new Vertex(outVId, (string)outVLabelList[0]!);

            // discard possible parent vertex
            await reader.ReadAsync(stream, cancellationToken).ConfigureAwait(false);

            var properties = await reader.ReadAsync(stream, cancellationToken).ConfigureAwait(false);
            var propertiesAsArray = null == properties ? Array.Empty<object>() : (properties as List<object>)?.ToArray();

            return new Edge(id, outV, label, inV, propertiesAsArray);
        }
    }
}

File: gremlin-js/gremlin-javascript/lib/structure/io/binary/internals/EdgeSerializer.js
/*
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */

/**
 * @author Igor Ostapenko
 */

import { Buffer } from 'buffer';
import { Edge, Vertex } from '../../../graph.js';

export default class EdgeSerializer {
  constructor(ioc) {
    this.ioc = ioc;
    this.ioc.serializers[ioc.DataType.EDGE] = this;
  }

  canBeUsedFor(value) {
    return value instanceof Edge;
  }

  serialize(item, fullyQualifiedFormat = true) {
    if (item === undefined || item === null) {
      if (fullyQualifiedFormat) {
        return Buffer.from([this.ioc.DataType.EDGE, 0x01]);
      }
      const id = [0x03, 0x00, 0x00, 0x00, 0x00, 0x00]; // String ''
      const label = [0x00, 0x00, 0x00, 0x00]; // empty list
      const inVId = [0x03, 0x00, 0x00, 0x00, 0x00, 0x00]; // String ''
      const inVLabel = [0x00, 0x00, 0x00, 0x00]; // empty list
      const outVId = [0x03, 0x00, 0x00, 0x00, 0x00, 0x00]; // String ''
      const outVLabel = [0x00, 0x00, 0x00, 0x00]; // empty list
      const parent = [0xfe, 0x01]; // null
      const properties = [0x09, 0x00, 0x00, 0x00, 0x00, 0x00]; // empty list
      return Buffer.from([...id, ...label, ...inVId, ...inVLabel, ...outVId, ...outVLabel, ...parent, ...properties]);
    }

    const bufs = [];
    if (fullyQualifiedFormat) {
      bufs.push(Buffer.from([this.ioc.DataType.EDGE, 0x00]));
    }

    // {id}
    bufs.push(this.ioc.anySerializer.serialize(item.id));

    // {label}
    const labels =
      item.labels instanceof Set
        ? Array.from(item.labels)
        : Array.isArray(item.label)
          ? item.label
          : item.label
            ? [item.label]
            : [];
    bufs.push(this.ioc.listSerializer.serialize(labels, false));

    // {inVId}
    const inVId = item.inV && item.inV.id;
    bufs.push(this.ioc.anySerializer.serialize(inVId));

    // {inVLabel}
    const inVLabels =
      item.inV && item.inV.labels instanceof Set
        ? Array.from(item.inV.labels)
        : item.inV && Array.isArray(item.inV.label)
          ? item.inV.label
          : item.inV && item.inV.label
            ? [item.inV.label]
            : [];
    bufs.push(this.ioc.listSerializer.serialize(inVLabels, false));

    // {outVId}
    const outVId = item.outV && item.outV.id;
    bufs.push(this.ioc.anySerializer.serialize(outVId));

    // {outVLabel}
    const outVLabels =
      item.outV && item.outV.labels instanceof Set
        ? Array.from(item.outV.labels)
        : item.outV && Array.isArray(item.outV.label)
          ? item.outV.label
          : item.outV && item.outV.label
            ? [item.outV.label]
            : [];
    bufs.push(this.ioc.listSerializer.serialize(outVLabels, false));

    // {parent}
    bufs.push(this.ioc.unspecifiedNullSerializer.serialize(null));

    // {properties}
    bufs.push(this.ioc.listSerializer.serialize([], true));

    return Buffer.concat(bufs);
  }

  /**
   * Async deserialization of edge value bytes from a StreamReader.
   * @param {StreamReader} reader
   * @param {number} valueFlag
   * @param {number} typeCode
   * @returns {Promise<Edge>}
   */
  async deserializeValue(reader, valueFlag, typeCode) {
    // {id} fully qualified
    const id = await this.ioc.anySerializer.deserialize(reader);

    // {label} bare list - full multi-label list
    const labelList = await this.ioc.listSerializer.deserializeValue(reader, 0x00, this.ioc.DataType.LIST);
    const labels = Array.isArray(labelList) ? labelList : labelList ? [labelList] : [];
    const label = labels.length > 0 ? labels[0] : 'edge';

    // {inVId} fully qualified
    const inVId = await this.ioc.anySerializer.deserialize(reader);

    // {inVLabel} bare list - full multi-label list
    const inVLabelList = await this.ioc.listSerializer.deserializeValue(reader, 0x00, this.ioc.DataType.LIST);
    const inVLabels = Array.isArray(inVLabelList) ? inVLabelList : inVLabelList ? [inVLabelList] : [];
    const inVLabel = inVLabels.length > 0 ? inVLabels[0] : 'vertex';

    // {outVId} fully qualified
    const outVId = await this.ioc.anySerializer.deserialize(reader);

    // {outVLabel} bare list - full multi-label list
    const outVLabelList = await this.ioc.listSerializer.deserializeValue(reader, 0x00, this.ioc.DataType.LIST);
    const outVLabels = Array.isArray(outVLabelList) ? outVLabelList : outVLabelList ? [outVLabelList] : [];
    const outVLabel = outVLabels.length > 0 ? outVLabels[0] : 'vertex';

    // {parent} fully qualified (always null in current TinkerPop)
    await this.ioc.anySerializer.deserialize(reader);

    // {properties} fully qualified
    const properties = await this.ioc.anySerializer.deserialize(reader);

    return new Edge(
      id,
      new Vertex(outVId, outVLabel, null, outVLabels),
      label,
      new Vertex(inVId, inVLabel, null, inVLabels),
      properties || [],
      labels,
    );
  }

  /**
   * Async fully-qualified deserialization from a StreamReader.
   * @param {StreamReader} reader
   * @returns {Promise<Edge|null>}
   */
  async deserialize(reader) {
    const type_code = await reader.readUInt8();
    if (type_code !== this.ioc.DataType.EDGE) {
      throw new Error(`EdgeSerializer: unexpected {type_code}=0x${type_code.toString(16)}`);
    }
    const value_flag = await reader.readUInt8();
    if (value_flag === 0x01) {
      return null;
    }
    if (value_flag !== 0x00) {
      throw new Error(`EdgeSerializer: unexpected {value_flag}=0x${value_flag.toString(16)}`);
    }
    return this.deserializeValue(reader, value_flag, type_code);
  }
}
